portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -15,27 +15,73 @@
 package com.google.sps.servlets;
 
 import java.util.ArrayList;
+import com.google.sps.data.Comment;
 import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+@WebServlet("/comment-data")
 public class DataServlet extends HttpServlet {
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json");
+        ArrayList<Comment> comments = new ArrayList<>();
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        int lim = getCommentLimit(request);
+        int cnt = 0;
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-
-    ArrayList<String> messages = new ArrayList<>();
-    messages.add("msg1");
-    messages.add("msg2");
-    messages.add("msg3");
-    Gson gson = new Gson();
-    String json = gson.toJson(messages);
-    response.getWriter().println(json);
-  }
+        Query query = new Query("Comment").addSort("time", SortDirection.DESCENDING);
+        PreparedQuery results = datastore.prepare(query);
+        for (Entity entity : results.asIterable()) {
+            String username = (String)entity.getProperty("username");
+            String comment = (String)entity.getProperty("comment");
+            long time = (long)entity.getProperty("time");
+            comments.add(new Comment(username, comment, System.currentTimeMillis()-time));
+            cnt++;
+            if(cnt == lim) break;
```

Comment:
`cnt` is zero based. Will this return one extra comment?

Reply:
No, cnt represents the number of comments we've added so far. Once it's equal it should break at the right place.

However I now realize this won't work if cnt=0, so I'll get that fixed.